homeassistant/components/backup/manager.py

```diff
@@ -0,0 +1,173 @@
+                            data = json.loads(data_file.read())
+                            backup = Backup(
+                                slug=data["slug"],
+                                name=data["name"],
+                                date=data["date"],
+                                path=backup_path,
+                                size=round(backup_path.stat().st_size / 1_048_576, 2),
+                            )
+                            backups[backup.slug] = backup
+                except (OSError, TarError, json.JSONDecodeError) as err:
+                    LOGGER.warning("Unable to read backup %s: %s", backup_path, err)
+
+        await self.hass.async_add_executor_job(_read_backups)
+        LOGGER.debug("Loaded %s backups", len(backups))
+        self._backups = backups
+        self._loaded = True
+
+    async def get_backups(self) -> dict[str, Backup]:
+        """Return backups."""
+        if not self._loaded:
+            await self.load_backups()
+
+        return self._backups
+
+    async def get_backup(self, slug: str) -> Backup | None:
+        """Return a backup."""
+        if not self._loaded:
+            await self.load_backups()
+
+        if not (backup := self._backups.get(slug)):
+            return None
+
+        if not backup.path.exists():
+            LOGGER.debug(
+                "Removing tracked backup (%s) that does not exists on the expected path %s",
+                backup.slug,
+                backup.path,
+            )
+            self._backups.pop(slug)
+            return None
+
+        return backup
+
+    async def remove_backup(self, slug: str) -> None:
+        """Remove a backup."""
+        if (backup := await self.get_backup(slug)) is None:
+            return
+
+        await self.hass.async_add_executor_job(backup.path.unlink, True)
+        LOGGER.debug("Removed backup located at %s", backup.path)
+        self._backups.pop(slug)
+
+    async def generate_backup(self) -> Backup:
+        """Generate a backup."""
+        if self.backing_up:
+            raise HomeAssistantError("Backup already in progress")
+
+        try:
+            self.backing_up = True
+            backup_name = f"Core {HAVERSION}"
+            date_str = dt.now().isoformat()
+            slug = _generate_slug(date_str, backup_name)
+
+            backup_data = {
+                "slug": slug,
+                "name": backup_name,
+                "date": date_str,
+                "type": "partial",
+                "folders": ["homeassistant"],
+                "homeassistant": {"version": HAVERSION},
+                "compressed": True,
+            }
+            tar_file_path = Path(self.backup_dir, f"{slug}.tar")
+
+            if not self.backup_dir.exists():
+                LOGGER.debug("Creating backup directory")
+                self.hass.async_add_executor_job(self.backup_dir.mkdir)
+
+            def _create_backup() -> None:
```

Comment:
Maybe make this a helper method that accepts the backup data, so we don't need to define the method every time we generate a backup?

Reply:
https://github.com/home-assistant/core/pull/67428

Reply:
This is not as wasteful as it seems. The code of this function is stored inside this code definition, so what it does is create a new "function" instance that reference the same code object each time.

```
>>> def create_helper():
...     def helper():
...         return 1
...     return helper
... 
>>> create_helper().__code__
<code object helper at 0x1029620e0, file "<stdin>", line 2>
>>> create_helper().__code__
<code object helper at 0x1029620e0, file "<stdin>", line 2>
```